testslide/__init__.py

```diff
@@ -115,6 +115,16 @@ def _all_methods(self):
     def _all_memoizable_attributes(self):
         return self._context.all_context_data_memoizable_attributes
 
+    def __setattr__(self, name, value):
+        if self.__dict__.get(name) and self.__dict__[name] != value:
+            raise AttributeError(
+                f"Attribute: {name} is already set.\
+        Resetting attributes in sub-contexts is not allowed, as it breaks composition.\
+            Use memoize and composition instead"
+            )
```

Comment:
This is a non-trivial thing to understand for the average users. I suggest giving a more thorough and detailed message here.

PS: I think the white spaces from code formatting were part of the string.

```
                f"Attribute {repr(name)} is already set.\n"
                "Changing the value of attributes after they have been set "
                "can lead to unexpected test results. Eg: when a sub context "
                "resets an attribute after a parent context has set and used "
                "it, they will have different objects for the same attribute.\n"
                "You can safely override attributes from parent contexts by "
                "using @context.before, @context.memoize_before or "
                "@context.function, so the inner-most definition is used."
```